libc/src/stdlib/qsort_r.cpp

```diff
@@ -19,13 +19,11 @@ LLVM_LIBC_FUNCTION(void, qsort_r,
                    (void *array, size_t array_size, size_t elem_size,
                     int (*compare)(const void *, const void *, void *),
                     void *arg)) {
-  if (array == nullptr || array_size == 0 || elem_size == 0)
-    return;
-  internal::Comparator c(compare, arg);
-  auto arr = internal::Array(reinterpret_cast<uint8_t *>(array), array_size,
-                             elem_size, c);
 
-  internal::sort(arr);
+  internal::unstable_sort(array, array_size, elem_size,
+                          [compare, arg](const void *a, const void *b) -> bool {
+                            return compare(a, b, arg) < 0;
+                          });
```

Comment:
@michaelrj-google I know this is ultimately subjective so I can't do better than an appeal to authority, but... as someone that knows a thing or two about sorting, the last decade of sorting theory has firmly discarded the ternary comparison in favor of a binary one. The public interface (obviously) can't be changed to match this, but the implementation however can be modern. I would strongly recommend to use the `is_less` paradigm in the implementation for consistency with all other modern sorting implementations, especially because the implemented sorting algorithm does not use the ternary nature of the comparison function at all.

Reply:
My concern is mainly around readability. I found it hard to determine what `is_less` did in the primary qsort files since it's always passed with a templated type, and I had a hard time finding where it came from because it's never declared as `is_less`, it's just an anonymous lambda. I don't think we can easily fix the template issue, since the lambdas for qsort and qsort_r have different types, but we might be able to solve the lack of a declaration.
```suggestion
  const auto is_less = [compare, arg](const void *a, const void *b) -> bool {
    return compare(a, b, arg) < 0;
  };
  internal::unstable_sort(array, array_size, elem_size, is_less);
```
This would make it much easier to find the source of `is_less`, without materially changing the code. Does this seem like a good solution to you?

Reply:
I appreciate your concern about readability and like the idea of pulling out the comparison function to a named variable, will do.